pkg/storage/tsdb/cacheable_index.go

```diff
@@ -0,0 +1,303 @@
+package tsdb
+
+import (
+	"bytes"
+	"context"
+	"encoding/gob"
+
+	"github.com/pkg/errors"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/model/labels"
+
+	"github.com/grafana/loki/pkg/storage/chunk/cache"
+	"github.com/grafana/loki/pkg/storage/tsdb/index"
+)
+
+const (
+	// sep defines the character used to separate the components of the cache key.
+	sep = "/"
+)
+
+// CacheableIndex is an implementation of the TSDB index that cache calls to all operations.
+type CacheableIndex struct {
+	// Index is an implementation of the TSDB index used to evaluate an operation that isn't cached.
+	Index
+
+	// Cache is a cache implementation used to store TSDB operation results.
+	cache.Cache
+}
+
+// NewCacheableIndex instantiate a new cacheable index given an index and a cache implementation.
+//
+// It assumes the cache is ready to be used.
+func NewCacheableIndex(index *TSDBIndex, cache cache.Cache) CacheableIndex {
+	return CacheableIndex{
+		Index: index,
+		Cache: cache,
+	}
+}
+
+// Stop does everything necessary to stop this index safely.
+func (i *CacheableIndex) Stop() error {
+	i.Cache.Stop()
+	return nil
+}
+
+// GetChunkRefs is a cached implementation of GetChunkRefs.
+//
+// It uses as a key for the cache all parameters (userID, from, through, shard, matchers) separated by a slash (/).
+func (i *CacheableIndex) GetChunkRefs(ctx context.Context, userID string, from, through model.Time, shard *index.ShardAnnotation, matchers ...*labels.Matcher) ([]ChunkRef, error) {
+	const opName = "GetChunkRefs"
+
+	mountKeyFn := func(userID string, from, through model.Time, shard *index.ShardAnnotation, matcher *labels.Matcher) string {
+		stringfiedShard := ""
+		if shard != nil {
+			stringfiedShard = shard.String()
+		}
+		if matcher == nil {
+			return userID + sep + from.String() + sep + through.String() + stringfiedShard + sep
+		}
+		return userID + sep + from.String() + sep + through.String() + stringfiedShard + sep + matcher.String()
+	}
```

Comment:
it looks like `mountKeyFn` is the same for each index call? we could reuse the same function defined once?